### ISSUE TITLE
Potential two different leaders on a same term

### ISSUE THREAD

**Issue by BusyJay** [tikv/tikv#3572]
Suppose a region A contains three peers a, b, c at term n. If c is uninitialized, a is stuck for some reason, then b can become leader at term n + 1. If c is restarted before accepting a snapshot, and a is recovered, then a can still start a campaign. Because we don't save hard state of an uninitialized peer, so c's term go back to 0. Hence a can win the campaign and become leader at term n + 1. So both a and b are the leaders of term n + 1, which violate the raft election safety.

However, such cases seems no harm except the violation itself. If no log is committed when b is leader, then the replica of b keeps the same as a. If there are logs committed, then a should be recovered or c has accepted a snapshot, in both cases, a can't become leader at term less than n + 1.

The root cause of the problem is that hard state is not saved when a peer is uninitialized, which introduce the risk of stepping back of term on the node.

**Comment by gengliqi** [comment]
It's the same as https://github.com/tikv/tikv/issues/8381, I think we must add some check before proposing `conf change` to forbid promoting an uninitialized learner to voter. /cc @NingLin-P

**Comment by NingLin-P** [comment]
> It's the same as #8381, I think we must add some check before proposing `conf change` to forbid promoting an uninitialized learner to voter. /cc @NingLin-P

This forbids is not enough, if PD adds a new peer as voter (it is allowed currently, and use by a lot of tests) this issue can still happen. Maybe an uninitialized voter should just ignore vote request, this is intuitive, since the voter can not persist its vote, it should not vote at all. But this also means a voter won't vote for election and log replication until it is initialized. /cc @gengliqi  @BusyJay

**Comment by gengliqi** [comment]
> > It's the same as #8381, I think we must add some check before proposing `conf change` to forbid promoting an uninitialized learner to voter. /cc @NingLin-P
> 
> This forbids is not enough, if PD adds a new peer as voter (it is allowed currently, and use by a lot of tests) this issue can still happen. Maybe an uninitialized voter should just ignore vote request, this is intuitive, since the voter can not persist its vote, it should not vote at all. But this also means a voter won't vote for election and log replication until it is initialized. /cc @gengliqi @BusyJay

It is enough.
If PD wants to directly add a new peer as a voter, the leader must prevent it.
For tests, I suggest changing all adding voters to 2 steps, i.e. adding a learner then promoting it to a voter.
If a voter can ignore vote request, the available issue will happen even when the majority of peers are alive.

**Comment by BusyJay** [comment]
Force using learner before adding voter LGTM, we have always followed the `learner -> voter` procedure for quite a long time.

> ..., it should not vote at all.

It may not be a good idea. For example, we may choose to unsafe remove all content of a corrupted store, which can make a voter back to uninitialized. I suggest not to add the limitation to allow easier recovery in strange case.

**Comment by NingLin-P** [comment]
> For example, we may choose to unsafe remove all content of a corrupted store, which can make a voter back to uninitialized.

Letting uninitialized voter (which not persist its vote) to vote is the problem here, preventing `confchange` add uninitialized voter may prevent this happend, but if it do happen (as the case @BusyJay mentioned) then the problem still exist.

**Comment by BusyJay** [comment]
Yes, we just need ensure that in all expected cases a voter should not become uninitialized again.
